pilot/pkg/networking/core/v1alpha3/route/route.go

```diff
@@ -735,3 +733,80 @@ func getHashPolicy(push *model.PushContext, dst *networking.DestinationWeight) *
 
 	return nil
 }
+
+type envoyRouteType int
+
+const (
+	envoyPath envoyRouteType = iota
+	envoyPrefix
+	envoyRegex
+	envoyCatchAll
+)
+
+// SortRoutes will reorder the routes in a virtual host
+// path before prefix
+// prefix before regex
+// regex before catchall prefix / or catch all regex *
+// longest path before shorter path
+// longest prefix before shorter prefix
+// longest regex before shorter regex. it doesn't make any difference, but
```

Comment:
For regex - I think it makes a difference, we should preserve the original order ( or not allow merging - i.e. just one virtual service should be allowed to have regex ).

I think the safest option remains to only allow one 'main' VirtualService to be merged with path+prefix only rules.